### PR TITLE
feature/dialogue-system

### PR DESCRIPTION
**Pull request by Banananaman91**
Begun creation of dialogue system, currently working on allowing selection responses for customised dialogue trees

### CODE REVIEW

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/Dialogue.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        [SerializeField] private int _npcID;
+        [SerializeField] private string _npcName;
+        [SerializeField] private List<Message> _messages;
+        public int NpcId => _npcID;
+        public String NpcName => _npcName;
+        public List<Message> Messages => _messages;
+    }
+
+    [Serializable]
+    public class Message
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/Dialogue.cs]: Move this into another file.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/Dialogue.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        [SerializeField] private int _npcMoodId;
+        [SerializeField] private string _messageText;
+        [SerializeField] private List<Response> _responses;
+        public int NpcMoodId => _npcMoodId;
+        public String MessageText => _messageText;
+        public List<Response> Responses => _responses;
+    }
+
+    [Serializable]
+    public class Response
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/Dialogue.cs]: Move this into another file.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/NpcImages.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+{
+    [CreateAssetMenu(fileName = "NpcImages", menuName = "NpcImages", order = 2)]
+    public class NpcImages : ScriptableObject
+    {
+        [SerializeField] private List<NpcMoods> _npcImage;
+        public List<NpcMoods> NpcImage => _npcImage;
+    }
+    
+    [Serializable]
+    public class NpcMoods
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/NpcImages.cs]: Move this into another file.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        [SerializeField] private Text _pageName;
+        [SerializeField] private Text _pageText;
+        [SerializeField] private GameObject _pageImagePosition;
+        [SerializeField] private float _sentenceSpeed;
+        [SerializeField] private Button _option;
+        [SerializeField] private GameObject[] _buttonPositions;
+        [SerializeField] private Image _dialogueBackground;
+        [SerializeField] private GameObject _dialogueBox;
+        private List<Button> _responseOptions = new List<Button>();
+        private Image previousImage;
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: This doesn't follow the codebase's naming convention for private fields.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+            int buttonCount = 0;
+            foreach (var response in npcResponses.Responses)
+            {
+                
+                var button = Instantiate(_option, _buttonPositions[buttonCount].transform.position, _buttonPositions[buttonCount].transform.rotation);
+                button.transform.SetParent(_dialogueBackground.transform);
+                button.GetComponentInChildren<Text>().text = response.Reply;
+                _responseOptions.Add(button);
+                button.onClick.AddListener(() => PlayParagraphCycle(npcMessage, npcImages, response.Next));
+                ++buttonCount;
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: why not just use a `for` loop here instead?

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+            }
+        }
+
+        private void EndDialogue()
+        {
+            _pageName.text = "";
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: please prefer `string.Empty` over `""` as intent is much clearer.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+            _pageText.text = "";
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: please prefer `string.Empty` over `""` as intent is much clearer.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/Dialogue.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace TurnBasedAssets.Scripts.Dialogue
+{
+    [CreateAssetMenu(fileName = "Dialogue", menuName = "Dialogue", order = 1)]
+    [Serializable]
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/Dialogue.cs]: Please combine these attributes into a single set of square brackets.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                sb.Append(letter);
+                RenderPageText(npc.NpcName, sb.ToString());
+                yield return new WaitForSeconds(_sentenceSpeed);
+            }
+            yield return null;
+        }
+
+        public void PlayParagraphCycle(Dialogue npcDialogue, NpcImages npcImages, int paragraphNumber)
+        {
+            if (_dialogueBox.activeSelf == false) _dialogueBox.SetActive(true);
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: prefer a NOT operator over a comparison to `false`.

**Review by RubyNova — CHANGES_REQUESTED** [review]

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs (hunk 5 above)
Comment (Banananaman91) [reply GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: Personal preference to use foreach like this, I find it clearer and nicer to use

Comment (RubyNova) [reply GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: I advise changing it still. This is precisely what `for` was meant for - not to mention you are introducing additional overhead by relying on `foreach`, albeit a tiny amount, its slightly noisy code.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/DialogueObject.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        [SerializeField] private NpcImages _npcImages;
+        [SerializeField] private RenderDialogue _pageRender;
+        [SerializeField] private int _startMessage;
+        
+        public void Select()
+        {
+            _pageRender.PlayParagraphCycle(_dialogue, _npcImages,_startMessage);
+        }
+
+        public void DeSelect()
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/DialogueObject.cs:20]: Is this supposed to be empty?

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+                foreach (var npcMood in _npcImageMoods.NpcMoodImages)
+                {
+                    var npcMoodName = npcMood.name.ToLower();
+                    if (npcMoodName.Contains(npcDialogue.Messages[paragraphNumber].NpcMood.ToLower()))
+                    {
+                        _newMoodImage = npcMood;
+                    }
+                }
+
+                if (!_dialogueBox.activeSelf)
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: You can one-line this if statement.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+                {
+                    _dialogueBox.SetActive(true);
+                }
+
+                if (IsPreviousImageNotNull)
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: You can one-line this if statement.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+                sb.Append(letter);
+                RenderPageText(npc.NpcName, sb.ToString());
+                yield return new WaitForSeconds(_sentenceSpeed);
+            }
+            yield return null;
+        }
+
+        public void PlayParagraphCycle(Dialogue npcDialogue, NpcImages npcImages, int paragraphNumber)
+        {
+            if (paragraphNumber < 0)
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs:48]: have you considered just `return`ing after calling `EndDialogue()` here? It'd reduce your scope count by one.

**Review by RubyNova — CHANGES_REQUESTED** [review]
Very minor things. Once these are dealt with I'll approve.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/DialogueObject.cs (hunk 10 above)
Comment (Banananaman91) [reply GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/DialogueObject.cs:20]: Yes, there was no functionality to add but the Interface required it to be there

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs (hunk 13 above)
Comment (Banananaman91) [reply GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs:48]: I don't understand?

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/DialogueObject.cs [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+namespace TurnBasedAssets.Scripts.Dialogue
+{
+    public class DialogueObject : MonoBehaviour, ISelection
+    {
+        [SerializeField] private Dialogue _dialogue;
+        [SerializeField] private NpcImages _npcImages;
+        [SerializeField] private RenderDialogue _pageRender;
+        [SerializeField] private int _startMessage;
+        
+        public void Select()
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/DialogueObject.cs]: You can expression body this.

File: GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+        {
+            var sb = new StringBuilder();
+            var letters = npcMessage.MessageText.ToCharArray();
+            foreach (var letter in letters)
+            {
+                sb.Append(letter);
+                RenderPageText(npc.NpcName, sb.ToString());
+                yield return new WaitForSeconds(_sentenceSpeed);
+            }
+            yield return null;
```
Comment (RubyNova) [inline GameProject/Assets/TurnBasedAssets/Scripts/Dialogue/RenderDialogue.cs]: This yield return null doesn't need to be here I don't think?

**Review by RubyNova — CHANGES_REQUESTED** [review]
Ok just these two tweaks and it should be good to go.

**Review by RubyNova — APPROVED** [review]
Good to go bar your scene conflict.

**Pull request merged**